api/views/share.py

```diff
@@ -92,6 +92,8 @@ def get_object(self):
 
         filter_kwargs = {self.lookup_field: decoded_pk}
         obj = get_object_or_404(queryset, **filter_kwargs)
+        while obj.same_as:
```

Comment:
Good call. I added a `get_canonical` method to ShareObjectManager, which uses a recursive query to follow same_as to the next unmerged object, and detect same_as cycles (not that they could ever happen...)